File: spring-boot/src/main/java/springboot/common/utility/JwtUtility.java
package springboot.common.utility;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.JwtParser;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.crypto.SecretKey;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import springboot.config.ConfigProvider;

@Component
public class JwtUtility {

    private final JwtParser jwtParser;
    private final SecretKey signKey;

    private final String userUid = "userUid";
    private final String userName = "userName";
    private final String jwtRoles;
    private final String securityRole;

    @Autowired
    public JwtUtility(ConfigProvider configProvider) {
        String rawSecretKey = configProvider.getJwtSecretKey();
        this.jwtRoles = configProvider.getJwtRole();
        this.securityRole = configProvider.getSecurityRole();
        this.signKey = Keys.hmacShaKeyFor(rawSecretKey.getBytes(StandardCharsets.UTF_8));
        this.jwtParser = Jwts.parserBuilder()
                             .setSigningKey(this.signKey)
                             .build();
    }

    public Claims jwtParse(String jwt) throws JwtException {
        return this.jwtParser.parseClaimsJws(jwt)
                             .getBody();
    }

    public String jwtSign(Long userUid, String userName) {
        return Jwts.builder()
                   .claim(this.userUid, userUid)
                   .claim(this.userName, userName)
                   .claim(this.securityRole, this.jwtRoles)
                   .signWith(this.signKey)
                   .compact();
    }

    public Long getUserUid(Claims claims) {
        return Long.parseLong((String) claims.get(this.userUid));
    }

    public String getUserName(Claims claims) {
        return (String) claims.get(this.userName);
    }
}

File: spring-boot-kt/src/main/kotlin/real/world/springbootkt/global/utility/JwtUtility.kt
package real.world.springbootkt.global.utility

import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtParser
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Component
import real.world.springbootkt.global.common.SecurityProperties
import java.nio.charset.StandardCharsets
import java.util.Date
import javax.crypto.SecretKey

@Component
class JwtUtility(securityProperties: SecurityProperties) {
    private val signKey: SecretKey = Keys.hmacShaKeyFor(securityProperties.jwtSecretKey.toByteArray(StandardCharsets.UTF_8))
    private val jwtParser: JwtParser = Jwts.parser().verifyWith(this.signKey).build()
    private val userId = "userId"
    private val userName = "userName"
    private val jwtRoles: String = securityProperties.jwtRole
    private val securityRole: String = securityProperties.securityRole

    fun jwtParse(jwt: String): Claims = jwtParser.parseSignedClaims(jwt).payload

    fun jwtSign(userId: Long, userName: String): String {
        return Jwts.builder()
            .claim(this.userId, userId)
            .claim(this.userName, userName)
            .claim(this.securityRole, this.jwtRoles)
            .signWith(this.signKey)
            .expiration(Date(System.currentTimeMillis() + 86400000))
            .compact()
    }
}